pkg/helpers/helpers.go

```diff
@@ -0,0 +1,641 @@
+		annotations = make(map[string]string)
+	}
+
+	currentVal, ok := annotations[key]
+	if !ok || currentVal != val {
+		annotations[key] = val
+		objMeta.SetAnnotations(annotations)
+		return true
+	}
+
+	return false
+}
+
+// GenerateClientFromSecret generate a client from a given secret
+func GenerateClientFromSecret(secret *corev1.Secret) (client.Client, meta.RESTMapper, error) {
+	var err error
+	var config *clientcmdapi.Config
+
+	if kubeconfig, ok := secret.Data["kubeconfig"]; ok {
+		config, err = clientcmd.Load(kubeconfig)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	token, tok := secret.Data["token"]
+	server, sok := secret.Data["server"]
+	if tok && sok {
+		config = clientcmdapi.NewConfig()
+		config.Clusters["default"] = &clientcmdapi.Cluster{
+			Server:                string(server),
+			InsecureSkipTLSVerify: true,
+		}
+		config.AuthInfos["default"] = &clientcmdapi.AuthInfo{
+			Token: string(token),
+		}
+		config.Contexts["default"] = &clientcmdapi.Context{
+			Cluster:  "default",
+			AuthInfo: "default",
+		}
+		config.CurrentContext = "default"
+	}
+
+	if config == nil {
+		return nil, nil, fmt.Errorf("kubeconfig or token and server are missing")
+	}
+
+	clientConfig, err := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	runtimeClient, err := client.New(clientConfig, client.Options{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(clientConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return runtimeClient, mapper, nil
+}
+
+// AddManagedClusterFinalizer add a finalizer to a managed cluster
+func AddManagedClusterFinalizer(client client.Client, recorder record.EventRecorder,
+	managedCluster *clusterv1.ManagedCluster, finalizer string) error {
+	for i := range managedCluster.Finalizers {
+		if managedCluster.Finalizers[i] == finalizer {
+			return nil
+		}
+	}
+
+	managedCluster.Finalizers = append(managedCluster.Finalizers, finalizer)
+	if err := client.Update(context.TODO(), managedCluster); err != nil {
+		return err
+	}
+
+	recorder.Eventf(managedCluster, "Normal", "ManagedClusterImportFinalizerAdded",
+		"The managed cluster %s finalizer %s is added", managedCluster.Name, finalizer)
+
+	return nil
+}
+
+// RemoveManagedClusterFinalizer remove a finalizer from a managed cluster
+func RemoveManagedClusterFinalizer(client client.Client, recorder record.EventRecorder,
+	managedCluster *clusterv1.ManagedCluster, finalizer string) error {
+	copiedFinalizers := []string{}
+	for i := range managedCluster.Finalizers {
+		if managedCluster.Finalizers[i] == finalizer {
+			continue
+		}
+		copiedFinalizers = append(copiedFinalizers, managedCluster.Finalizers[i])
+	}
+
+	if len(managedCluster.Finalizers) == len(copiedFinalizers) {
+		return nil
+	}
+
+	managedCluster.Finalizers = copiedFinalizers
+	if err := client.Update(context.TODO(), managedCluster); err != nil {
```

Comment:
should we consider retry here if fails? or write another update function with retry?

Reply:
retry here may be not necessary, if this method return error, the controller will reconcile the request, it will equal retry